### PR TITLE
MINOR: Refactor SslFactory

### PR DESCRIPTION
**Pull request by cmccabe**
SslFactory: split the part of SslFactory that creates SSLEngine instances into SslEngineBuilder.  When (re)configuring, we simply create a new SslEngineBuilder.  This allows us to make all the builder fields immutable.  It also simplifies the logic for reconfiguring.  Because we   sometimes need to test old SslEngine instances against new ones, being able to use both the old and the new builder at once is useful.

Create an enum named SslClientAuth which encodes the possible values for ssl.client.auth.  This will simplify the handling of this configuration.

SslTransportLayer#maybeProcessHandshakeFailure should treat an SSLHandshakeException with a "Received fatal alert" message as a handshake error (and therefore an authentication error.)

SslFactoryTest: add some line breaks for very long lines.

ConfigCommand#main: when terminating the command due to an uncaught exception, log the exception using debug level in slf4j, in addition to printing it to stderr.  This makes it easier to debug failing junit tests, where stderr may not be kept, or may be reordered with respect to  other slf4j messages.  The use of debug level is consistent with how we handle other types of exceptions in ConfigCommand#main.

StateChangeLogMerger#main: spell out the full name of scala.io.Source rather than abbreviating it as io.Source.  This makes it clearer that it is part of the Scala standard library.  It also avoids compiler errors when other libraries whose groupId starts with "io" are used in the broker.

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/config/SslClientAuth.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ */
+public enum SslClientAuth {
+    REQUIRED,
+    REQUESTED,
+    NONE;
+
+    public static final List<SslClientAuth> VALUES =
+        Collections.unmodifiableList(Arrays.asList(SslClientAuth.values()));
+
+    public static SslClientAuth create(String key) {
```
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/common/config/SslClientAuth.java]: I think we tend to use `forXxx` rather than `create` for mapping names to enums.

File: clients/src/main/java/org/apache/kafka/common/config/SslClientAuth.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        if (key == null) {
+            return SslClientAuth.NONE;
+        }
+        key = key.toLowerCase(Locale.ROOT);
```
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/common/config/SslClientAuth.java]: Better to do `toUpperCase` to avoid the `toLowerCase` in the loop?

File: clients/src/main/java/org/apache/kafka/common/config/SslConfigs.java [hunk 3]
```diff
@@ -96,7 +95,7 @@
     public static final String SSL_KEYMANAGER_ALGORITHM_CONFIG = "ssl.keymanager.algorithm";
     public static final String SSL_KEYMANAGER_ALGORITHM_DOC = "The algorithm used by key manager factory for SSL connections. "
             + "Default value is the key manager factory algorithm configured for the Java Virtual Machine.";
-    public static final String DEFAULT_SSL_KEYMANGER_ALGORITHM = KeyManagerFactory.getDefaultAlgorithm();
+    public static final String DEFAULT_SSL_KEYMANGER_ALGORITHM = null;
```
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/common/config/SslConfigs.java]: What is the rationale for this change? It will be better to handle defaults consistently for keymanager algorithm and trustmanager algorithm. And since the doc says we use the JVM default, it seems to make sense to set the default here.

File: clients/src/test/java/org/apache/kafka/common/security/ssl/SslFactoryTest.java [hunk 4]
```diff
@@ -48,18 +48,19 @@
...
         SslFactory sslFactory = new SslFactory(Mode.SERVER);
         sslFactory.configure(serverSslConfig);
         //host and port are hints
         SSLEngine engine = sslFactory.createSslEngine("localhost", 0);
         assertNotNull(engine);
-        String[] expectedProtocols = {"TLSv1.2"};
-        assertArrayEquals(expectedProtocols, engine.getEnabledProtocols());
+        HashSet<String> expectedProtocols = new HashSet<>();
+        expectedProtocols.add("TLSv1.2");
+        assertEquals(expectedProtocols, new HashSet<>(Arrays.asList(engine.getEnabledProtocols())));
```
Comment (rajinisivaram) [inline clients/src/test/java/org/apache/kafka/common/security/ssl/SslFactoryTest.java]: Can replace the three lines with:
```
assertEquals(Utils.mkSet("TLSv1.2"), Utils.mkSet(engine.getEnabledProtocols()));
```

File: clients/src/test/java/org/apache/kafka/common/security/ssl/SslFactoryTest.java [hunk 5]
```diff
@@ -106,79 +107,78 @@ public void testClientMode() throws Exception {
...
-        assertFalse(sslFactory.createSslEngine("localhost", 0).getUseClientMode());
+        SslEngineBuilder sslEngineBuilder = sslFactory.sslEngineBuilder();
+        assertNotNull("SslEngineBuilder not created", sslEngineBuilder);
 
-        // Verify that context is not recreated on reconfigure() if config and file are not changed
+        // Verify that SslEngineBuilder is not recreated on reconfigure() if config and
+        // file are not changed
         sslFactory.reconfigure(sslConfig);
-        assertSame("SSL context recreated unnecessarily", sslContext, sslFactory.sslContext());
+        assertSame("SSL context recreated unnecessarily",
```
Comment (rajinisivaram) [inline clients/src/test/java/org/apache/kafka/common/security/ssl/SslFactoryTest.java]: `SSL context` => `SslEngineBuilder`

File: clients/src/test/java/org/apache/kafka/common/security/ssl/SslFactoryTest.java [hunk 6]
```diff
@@ -106,79 +107,78 @@ public void testClientMode() throws Exception {
...
+                sslFactory.sslEngineBuilder().sslContext());
         assertFalse(sslFactory.createSslEngine("localhost", 0).getUseClientMode());
 
-        trustStoreFile = File.createTempFile("truststore", ".jks");
-        sslConfig = TestSslUtils.createSslConfig(false, true, Mode.SERVER, trustStoreFile, "server");
-        sslConfig.remove(SslConfigs.SSL_TRUSTSTORE_LOCATION_CONFIG);
-        sslConfig.remove(SslConfigs.SSL_TRUSTSTORE_PASSWORD_CONFIG);
-        sslConfig.remove(SslConfigs.SSL_TRUSTSTORE_TYPE_CONFIG);
-        sslFactory.reconfigure(sslConfig);
-        assertNotSame("SSL context not recreated", sslContext, sslFactory.sslContext());
```
Comment (rajinisivaram) [inline clients/src/test/java/org/apache/kafka/common/security/ssl/SslFactoryTest.java:176]: Why was this removed? This is testing reconfiguration without a truststore.

File: clients/src/main/java/org/apache/kafka/common/security/ssl/SslEngineBuilder.java [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+
+    private static SslClientAuth createSslClientAuth(String key) {
+        SslClientAuth auth = SslClientAuth.create(key);
+        if (auth != null) {
+            return auth;
+        }
+        log.warn("Unrecognized client authentication configuration {}.  Falling " +
+                "back to NONE.  Recognized client authentication configurations are {}.",
+                key, String.join(", ", SslClientAuth.VALUES.stream().
+                        map(a -> a.name()).collect(Collectors.toList())));
```
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/common/security/ssl/SslEngineBuilder.java:116]: Shouldn't we throw an exception instead?

File: clients/src/main/java/org/apache/kafka/common/security/ssl/SslEngineBuilder.java [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+     * @param peerPort  The peer port to use.
+     * @return          The new SSLEngine.
+     */
+    public SSLEngine createSslEngine(Mode mode, String peerHost, int peerPort) {
+        SSLEngine sslEngine = sslContext.createSSLEngine(peerHost, peerPort);
+        if (cipherSuites != null) sslEngine.setEnabledCipherSuites(cipherSuites);
+        if (enabledProtocols != null) sslEngine.setEnabledProtocols(enabledProtocols);
+
+        // SSLParameters#setEndpointIdentificationAlgorithm enables endpoint validation
+        // only in client mode. Hence, validation is enabled only for clients.
```
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/common/security/ssl/SslEngineBuilder.java]: Move this comment to the else block?

File: clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java [hunk 9]
```diff
@@ -63,86 +52,50 @@
...
-            this.endpointIdentification = endpointIdentification;
-
-        String secureRandomImplementation = (String) configs.get(SslConfigs.SSL_SECURE_RANDOM_IMPLEMENTATION_CONFIG);
-        if (secureRandomImplementation != null) {
-            try {
-                this.secureRandomImplementation = SecureRandom.getInstance(secureRandomImplementation);
-            } catch (GeneralSecurityException e) {
-                throw new KafkaException(e);
-            }
+    public void configure(Map<String, ?> newConfigs) throws KafkaException {
```
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java]: `configs` instead of `newConfigs` since this is the initial configuration?

File: clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java [hunk 10]
```diff
@@ -151,236 +104,142 @@ else if (clientAuthConfig.equals("requested"))
...
     @Override
-    public void validateReconfiguration(Map<String, ?> configs) {
+    public void validateReconfiguration(Map<String, ?> newConfigs) {
+        if (sslEngineBuilder == null) {
+            throw new IllegalStateException("SslFactory has not been configured.");
+        }
+        Map<String, Object> nextConfigs = new HashMap<>(sslEngineBuilder.configs());
+        copyMapEntries(nextConfigs, newConfigs, SslConfigs.RECONFIGURABLE_CONFIGS);
+        if (clientAuthConfigOverride != null) {
+            nextConfigs.put(BrokerSecurityConfigs.SSL_CLIENT_AUTH_CONFIG, clientAuthConfigOverride);
```
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java:123]: Why does this need special handling?

File: clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java [hunk 11]
```diff
@@ -151,236 +104,142 @@ else if (clientAuthConfig.equals("requested"))
...
+            }
+            if (keystoreVerifiableUsingTruststore) {
+                if (sslEngineBuilder.truststore() != null || sslEngineBuilder.keystore() != null) {
+                    SslEngineValidator.validate(sslEngineBuilder, newSslEngineBuilder);
+                }
             }
         } catch (Exception e) {
-            log.debug("Validation of dynamic config update of SSL keystore/truststore failed", e);
-            throw new ConfigException("Validation of dynamic config update of SSL keystore/truststore failed: " + e);
+            log.info("Unable to validate proposed new SslFactory configuration.", e);
```
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java]: Since this is a client-side error propagated to the client, we should only log at debug level?

File: clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java [hunk 12]
```diff
@@ -151,236 +104,142 @@ else if (clientAuthConfig.equals("requested"))
...
+            throw new ConfigException("Unable to reconfigure SslFactory: " + e);
```
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java]: This is a validation failure (e.g. admin client validateOnly request), sobetter to say `validation failed`.

File: clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java [hunk 13]
```diff
@@ -151,236 +104,142 @@ else if (clientAuthConfig.equals("requested"))
...
-                    !Objects.equals(configs.get(SslConfigs.SSL_KEYSTORE_LOCATION_CONFIG), keystore.path) ||
-                    !Objects.equals(configs.get(SslConfigs.SSL_KEYSTORE_PASSWORD_CONFIG), keystore.password) ||
-                    !Objects.equals(configs.get(SslConfigs.SSL_KEY_PASSWORD_CONFIG), keystore.keyPassword);
-            if (!keystoreChanged) {
-                keystoreChanged = keystore.modified();
-            }
+        Map<String, Object> nextConfigs = new HashMap<>(sslEngineBuilder.configs());
+        copyMapEntries(nextConfigs, newConfigs, SslConfigs.RECONFIGURABLE_CONFIGS);
+        if (clientAuthConfigOverride != null) {
+            nextConfigs.put(BrokerSecurityConfigs.SSL_CLIENT_AUTH_CONFIG, clientAuthConfigOverride);
```
Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java]: Same as in validate, why the special handling? Isn't this included under `sslEngineBuilder.configs()`?

File: clients/src/test/java/org/apache/kafka/common/network/SslTransportLayerTest.java [hunk 14]
```diff
@@ -1014,12 +1014,6 @@ private void verifyInvalidReconfigure(ListenerReconfigurable reconfigurable,
         } catch (KafkaException e) {
             // expected exception
         }
-        try {
-            reconfigurable.reconfigure(invalidConfigs);
```
Comment (rajinisivaram) [inline clients/src/test/java/org/apache/kafka/common/network/SslTransportLayerTest.java]: We used to guarantee that reconfigure fails if configs are invalid. Is that no longer the case?

**Review by rajinisivaram — COMMENTED** [review]
@cmccabe Thanks for the PR, looks good overall. Left some comments.

Can we make sure that the changes here are consistent with the proposal in https://cwiki.apache.org/confluence/display/KAFKA/KIP-383%3A++Pluggable+interface+for+SSL+Factory?

We have definitions of client auth in KafkaConfig, we probably want to use the new enum there as well.

There seem to be a lot of changes to tests for which it is hard to tell why the changes were made. Do you have a list of changes or comments that could be useful in reviewing these?

File: clients/src/main/java/org/apache/kafka/common/config/SslClientAuth.java (hunk 1 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/common/config/SslClientAuth.java]: OK.  I changed it to `forConfig` to be more consistent with that.

File: clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java (hunk 10 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java:123]: As you know, `clientAuthConfigOverride` functions as an override.  If it is set, the value of `SSL_CLIENT_AUTH_CONFIG` given by the user is irrelevant.  It's simplest just to overwrite the value given by the user.

Of course, since `SSL_CLIENT_AUTH_CONFIG` isn't currently reconfigurable, this isn't an issue in practice right now.  But, perhaps we will let this be reconfigured in the future.

File: clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java (hunk 11 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java]: ok

File: clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java (hunk 13 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java]: same issue as above -- the override should overwrite whatever the user supplied

File: clients/src/test/java/org/apache/kafka/common/network/SslTransportLayerTest.java (hunk 14 above)
Comment (cmccabe) [reply clients/src/test/java/org/apache/kafka/common/network/SslTransportLayerTest.java]: It seems wasteful to do validation twice.  `validateReconfiguration` should be called before `reconfigure`.  Then `reconfigure` doesn't have to repeat all of the checks that were done in `validateReconfiguration`.

File: clients/src/main/java/org/apache/kafka/common/config/SslConfigs.java (hunk 3 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/common/config/SslConfigs.java]: I'll revert the change to this constant

**Comment by cmccabe** [comment]
> Can we make sure that the changes here are consistent with the proposal in https://cwiki.apache.org/confluence/display/KAFKA/KIP-383%3A++Pluggable+interface+for+SSL+Factory?

I think they are consistent.  I think `SslEngineBuilder` is pretty similar to what KIP-383 calls `SslEngineFactory`.  The KIP makes the point that "the idea is to make the configuration immutable in the pluggable factory"-- exactly what this PR does with `SslEngineBuilder`. 
 `SslEngineFactory` is also supposed to have only one method: `SSLEngine createSslEngine(String peerHost, int peerPort);`.  This is very similar to `SslEngineBuilder`, which mainly has `SSLEngine createSslEngine(Mode mode, String peerHost, int peerPort)` Since the KIP hasn't been accepted yet, there is still time to modify it or think it over.  But overall it seems compatible.

> We have definitions of client auth in KafkaConfig, we probably want to use the new enum there as well.

Good point.  I added the new enum to {{KafkaConfig.scala}}.

> There seem to be a lot of changes to tests for which it is hard to tell why the changes were made. Do you have a list of changes or comments that could be useful in reviewing these?

* `SslTransportLayerTest.java`: we no longer test that `SslFactory#reconfigure` validates that the reconfiguration can be done.  The reason is because validation only happens in `SslFactory#validateReconfiguration`, rather than being duplicated in both functions.

* `SslFactoryTest.java`: mainly, the changes here are to test `SslFactory#sslEngineBuilder` rather than `SslFactory#sslContext`.  For example, we check that the engine builder is re-created on re-configuration, rather than checking that the ssl context is re-created.  The rationale here is that SSL context is somewhat of an implementation detail and it would be better to not poke at it unless we have a reason.  This will help if we someday have another way of creating SSL engines besides JSSE, etc.  There is also a change where we compare sets rather than comparing arrays.  The rationale here is that in the collection we're looking at, ordering doesn't matter.

File: clients/src/test/java/org/apache/kafka/common/network/SslTransportLayerTest.java (hunk 14 above)
Comment (rajinisivaram) [reply clients/src/test/java/org/apache/kafka/common/network/SslTransportLayerTest.java]: The file may be modified asynchronously and the keystore and truststore loaded from file during `reconfigure` may not be the same as the one used during `validateReconfiguration`. Unless we guarantee that we save and use the stores loaded during validation, we have to ensure that `reconfigure` doesn't break the broker.

**Comment by rajinisivaram** [comment]
@cmccabe Thanks for the updates, PR builds are failing, can you take a look? Also left a comment on validation.

File: clients/src/test/java/org/apache/kafka/common/security/ssl/SslFactoryTest.java (hunk 6 above)
Comment (cmccabe) [reply clients/src/test/java/org/apache/kafka/common/security/ssl/SslFactoryTest.java:176]: this is another example of the validate / reconfigure split: reconfigure no longer re-performs all the work validation does

File: clients/src/main/java/org/apache/kafka/common/security/ssl/SslEngineBuilder.java (hunk 7 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/common/security/ssl/SslEngineBuilder.java:116]: The existing behavior is to fall back on NONE when something malformed or unknown is supplied.  Changing that would be a breaking change since some brokers might not start up after a rolling upgrade.

File: clients/src/main/java/org/apache/kafka/common/security/ssl/SslEngineBuilder.java (hunk 8 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/common/security/ssl/SslEngineBuilder.java]: good idea

File: clients/src/test/java/org/apache/kafka/common/network/SslTransportLayerTest.java (hunk 14 above)
Comment (cmccabe) [reply clients/src/test/java/org/apache/kafka/common/network/SslTransportLayerTest.java]: Even if `reconfigure` re-does the validation, the file could still be changed in between the validation and when we start using it.  There is no way to fully close the race condition except by copying the file to a separate location that only we control.  If we want to do that, it should probably be a separate follow-on change.

In general, I don't think modifying the truststore and keystore files in place will ever work very well.  After all, when those files are modified in place, there will always be some short time period the modification and Kafka re-validating that the configuration still works.  And there will also be no working configuration to fall back on in that case, since the original file is gone.

Comment (rajinisivaram) [reply clients/src/test/java/org/apache/kafka/common/network/SslTransportLayerTest.java]: @cmccabe `reconfigure` was testing that the keystores and truststores loaded from files by `reconfigure` and used to create the new SSL context are actually valid. If the file changes after that, it is fine since the new files will not be used for creating a new context until the next admin client request to update. For this particular invocation of `reconfigure`, we need to guarantee that SSLContext is updated only if the keystore/truststore used to the create the new SSLContext are valid.

Even though we tried to push back on in-place updates of keystore/truststore, it seems to be what everyone wants to do. And hence supporting this without any timing issues is critical.

**Comment by cmccabe** [comment]
OK.  I revised the patch so that `reconfigure` also does validation.

File: clients/src/test/java/org/apache/kafka/common/network/SslTransportLayerTest.java (hunk 14 above)
Comment (rajinisivaram) [reply clients/src/test/java/org/apache/kafka/common/network/SslTransportLayerTest.java]: @cmccabe Can we keep this test now since we validate on reconfiguration?

**Review by rajinisivaram — APPROVED** [review]
@cmccabe Thanks for the updates. Can we revert the changes to `SslTransportLayerTest.java` since we validate configs during reconfiguration? Apart from that LGTM.

File: clients/src/test/java/org/apache/kafka/common/network/SslTransportLayerTest.java (hunk 14 above)
Comment (cmccabe) [reply clients/src/test/java/org/apache/kafka/common/network/SslTransportLayerTest.java]: I changed it back.

**Comment by cmccabe** [comment]
> Can we revert the changes to SslTransportLayerTest.java since we validate configs during reconfiguration

Yeah, reverted.

Thanks for the reviews.  I'll push a new version and get a build from Jenkins :)
